pkg/controller/kubelet-config/kubelet_config_controller.go

```diff
@@ -0,0 +1,371 @@
+		return
+	}
+	ctrl.queue.Add(key)
+}
+
+func (ctrl *Controller) enqueueRateLimited(cfg *mcfgv1.KubeletConfig) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(cfg)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", cfg, err))
+		return
+	}
+	ctrl.queue.AddRateLimited(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (ctrl *Controller) worker() {
+	for ctrl.processNextWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextWorkItem() bool {
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	err := ctrl.syncHandler(key.(string))
+	ctrl.handleErr(err, key)
+
+	return true
+}
+
+func (ctrl *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.queue.Forget(key)
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing kubeletconfig %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(2).Infof("Dropping kubeletconfig %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+func findKubeletConfig(mc *mcfgv1.MachineConfig) (int, error) {
+	for i, c := range mc.Spec.Config.Storage.Files {
+		if c.Path == "/etc/kubernetes/kubelet.conf" {
+			return i, nil
+		}
+	}
+	return -1, fmt.Errorf("Could not find Kubelet Config")
+}
+
+// syncKubeletConfig will sync the kubeletconfig with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (ctrl *Controller) syncKubeletConfig(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing kubeletconfig %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing kubeletconfig %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	newKubeletConfig, err := ctrl.mckLister.Get(name)
+	if errors.IsNotFound(err) {
+		glog.V(2).Infof("KubeletConfig %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Get MachineConfig
+	mc, err := ctrl.client.Machineconfiguration().MachineConfigs().Get(newKubeletConfig.Spec.MachineConfigSelector, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// find the kubelet.conf
+	kubeletConfigIndex, err := findKubeletConfig(mc)
+	if err != nil {
+		return err
+	}
+	// Found the kubelet.conf
+	c := mc.Spec.Config.Storage.Files[kubeletConfigIndex]
+	dataURL, err := dataurl.DecodeString(c.Contents.Source)
+	if err != nil {
+		return err
+	}
+	originalKubeConfig, err := decodeKubeletConfig(dataURL.Data)
+	if err != nil {
+		return err
+	}
+	err = mergo.Merge(originalKubeConfig, newKubeletConfig.Spec.KubeletConfig, mergo.WithOverride)
```

Comment:
would be nice to see if there is something already in the vendor tree that can do this vs importing a new dep for this one call

Reply:
actually, this is already in the vendor tree. nm.